File: Core/src/main/java/jokatu/game/status/StandardTextStatus.java
package jokatu.game.status;

import com.fasterxml.jackson.annotation.JsonValue;
import jokatu.game.event.StatusUpdateEvent;
import ophelia.event.observable.AbstractSynchronousObservable;
import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;

/**
 * @author steven
 */
public class StandardTextStatus extends AbstractSynchronousObservable<StatusUpdateEvent> implements Status {

	private String text;

	public StandardTextStatus() {}

	public StandardTextStatus(String text) {
		this.text = text;
	}

	@NotNull
	@JsonValue
	@Override
	public String getText() {
		return text;
	}

	public void setText(String text, Object... arguments) {
		this.text = MessageFormat.format(text, arguments);
		fireEvent(() -> StandardTextStatus.this);
	}
}

File: Core/src/main/kotlin/jokatu/game/status/StandardTextStatus.kt
package jokatu.game.status

import com.fasterxml.jackson.annotation.JsonValue
import jokatu.game.event.StandardStatusUpdateEvent
import jokatu.game.event.StatusUpdateEvent
import ophelia.event.observable.AbstractSynchronousObservable

/**
 * A [Status] that fires a [StatusUpdateEvent] whenever its text is updated.
 * @author steven
 */
class StandardTextStatus : AbstractSynchronousObservable<StatusUpdateEvent>(), Status {

	@get:JsonValue
	override var text: String? = null
		set(value) {
			field = value
			fireEvent(StandardStatusUpdateEvent(this))
		}
}
